contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -0,0 +1,903 @@
+#include "contrib/golang/filters/http/source/golang_filter.h"
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "envoy/http/codes.h"
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/base64.h"
+#include "source/common/common/enum_to_int.h"
+#include "source/common/common/lock_guard.h"
+#include "source/common/common/utility.h"
+#include "source/common/grpc/common.h"
+#include "source/common/grpc/context_impl.h"
+#include "source/common/grpc/status.h"
+#include "source/common/http/headers.h"
+#include "source/common/http/http1/codec_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Golang {
+
+void Filter::onHeadersModified() {
+  // Any changes to request headers can affect how the request is going to be
+  // routed. If we are changing the headers we also need to clear the route
+  // cache.
+  decoding_state_.getFilterCallbacks()->downstreamCallbacks()->clearRouteCache();
+}
+
+Http::LocalErrorStatus Filter::onLocalReply(const LocalReplyData& data) {
+  auto& state = getProcessorState();
+  ASSERT(state.isThreadSafe());
+  ENVOY_LOG(debug, "golang filter onLocalReply, state: {}, phase: {}, code: {}", state.stateStr(),
+            state.phaseStr(), int(data.code_));
+
+  // TODO: let the running go filter return a bit earilier, by setting a flag?
+  return Http::LocalErrorStatus::Continue;
+}
+
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
+  ProcessorState& state = decoding_state_;
+
+  ENVOY_LOG(debug, "golang filter decodeHeaders, state: {}, phase: {}, end_stream: {}",
+            state.stateStr(), state.phaseStr(), end_stream);
+
+  if (dynamic_lib_ == nullptr) {
+    ENVOY_LOG(error, "dynamic_lib_ is nullPtr, maybe the instance already unpub.");
```

Comment:
How can this happen? Shouldn't you fail to load the filter and once it's loaded the DSO should remain loaded? I think this should all be removed.

Reply:
Done. yeah, merge the DSO to filter self and throw error at the config parsing phase when load failed. Thanks.